README.md
# IBMSETI

This README describes the `ibmseti` package. This software is used to read and analyze 
SETI Institute data generated from the Allen Telescope Array. It provides the essential 
code needed to get started reading the data, calculating spectrograms and autocorrelation
spectra, and extracting some of the standard features from those results.


## Example User Code

### Installation

    pip install ibmseti


### Read the Data

The raw data (`compamp` or `archive-compamp` files) are read with a `ibmseti.compamp.Compamp` object.
Be sure to open files in `rb` mode. 

```python
import ibmseti
rawdatafile = open('path/to/data/2014-08-11/act17713/2014-08-11_05-50-10_UTC.act17713.dx2009.id-2.R.archive-compamp','rb')
aca = ibmseti.compamp.Compamp(rawdatafile.read())
```

#### Simulated data

This package can now read data that was simulated for the [SETI Institute hackathon](https://www.eventbrite.com/e/seti-institute-hackathon-machine-learning-for-the-search-for-extraterrestrial-intelligence-tickets-32830428696) and [code challenge (June/July 2017)](https://www.eventbrite.com/e/seti-institute-code-challenge-machine-learning-for-the-search-for-extraterrestrial-intelligence-registration-32831985352)

```python
import ibmseti
rawdatafile = open('path/to/data/<some random string>.dat','rb')
aca = ibmseti.compamp.SimCompamp(rawdatafile.read())
```


### Create Spectrogram and Plot

Here, we show the code to produce a Spectrogram plot from a Compamp object with the `ibmseti.dsp` module.

```python
spectrogram = ibmseti.dsp.compamp_to_spectrogram(aca)
time_bins = ibmseti.dsp.time_bins( aca.header() )
freq_bins = ibmseti.dsp.frequency_bins( aca.header() )

import matplotlib.pyplot as plt
plt.ion()
fig, ax = plt.subplots()
ax.pcolormesh(freq_bins, time_bins, spectrogram)
```

#### Spectrum from Simulated Data

Due to the difference in the structure of the bytes packed in the simulated data compared to 
the real data (the simulated data are much simpler), a different code was written to create a spectrogram.
Also, there is no need for the time and frequency bin edges for the plots. The simulated signals could
be at any central frequency, and digitized at any sampling rate. However, the shape of the spectrogram 
produced in the code below transforms the data into the "standard" spectrogram/waterfall typically used
to analyze the real archive-compamp files.  Be sure to open files in `rb` mode. 

```python
rawdatafile = open('path/to/data/<uuid>.dat','rb')
aca = ibmseti.compamp.SimCompamp(rawdatafile.read())
spectrogram = aca.get_spectrogram()

import matplotlib.pyplot as plt
plt.ion()
fig, ax = plt.subplots()
ax.imshow(np.log(spectrogram), aspect = 0.5*float(spectrogram.shape[1]) / spectrogram.shape[0])
```


### Feature Extraction

Here we show some features one may extract from a spectrogram with the `ibmset.features` module.
In each of these, the `data` is the spectrogram.

It should be noted that these features are experimental and may or may not have strong 
impacts on classification. They are initial guesses made by a handful of SETI/NASA researchers, who
make no guarantees about their usefulness. But we add them in the code here to get you started. 

Also, the spectrograms are relatively large. One may consider reducing the size of the spectrogram
using https://gist.github.com/derricw/95eab740e1b08b78c03f. Calculations done with a 
spectogram of reduced size can be significantly faster. For example, one can reduce the
size from 129x6144 to 43x192: `spectrogram = bin_ndarray(spectrogram, (43,192), operation='sum')`). If you sliced
off one of the time-bins to make the spectrogram 128x6144, you could reduce the spectrogram by other factors.

##### Features based on the Spectrogram

###### Standard Deviation of the projection of the spectrogram onto the time and frequency axis.

```
std_time = math.sqrt(ibmseti.features.moment( ibmseti.features.projection(data, axis=1), moment=2))
std_freq = math.sqrt(ibmseti.features.moment( ibmseti.features.projection(data, axis=0), moment=2))
```

###### Average Standard Deviation of the spectrogram along each slice of the time and frequency axis.

```
std_time = np.mean(np.sqrt(ibmseti.features.moment( data, axis=0, moment=2)))
std_freq = np.mean(np.sqrt(ibmseti.features.moment( data, axis=1, moment=2)))
```

###### Average N-th moment of the spectrogram along each slice of the time and frequency axis.

```
N = 3 #kurtosis 
N = 4 #skewness
nth_moment_time = np.mean( ibmseti.features.moment( data, axis=0, moment=N))
nth_moment_freq = np.mean( ibmseti.features.moment( data, axis=1, moment=N))
```

###### N-th moment of the projection of the spectrogram onto the time and frequency axis.

```
N = 3 #kurtosis
N = 4 #skewness
nth_moment_time = ibmseti.features.moment( ibmseti.features.projection(data, axis=1), moment=N)
nth_moment_freq = ibmseti.features.moment( ibmseti.features.projection(data, axis=0), moment=N)
```

###### Mean [Total Variation](https://en.wikipedia.org/wiki/Total_variation) (along the time axis) of the spectrogram.

```
tv_for_each_frequency = ibmseti.features.total_variation(data)
tv = np.mean( tv_for_each_frequency )
```

###### Excess Kurtosis

Measures the *non-guassianity* of a distribution. Could be measure after projecting a spectrogram onto
it's frequency-axis.

```
fourth_mom = ibmseti.features.moment( ibmseti.features.projection(data, axis=0), moment=4)
variance = ibmseti.features.moment( ibmseti.features.projection(data, axis=0), moment=2)
excess_kurtosis = fourth_mom/variance - 3
```

##### Entropy (based on histogram of log of power)

The entropy of a signal is a measure of the amount of order/disorder in the system. In information
theory, it is the measure of the average amount of "information" in a signal. 

The `ibmseti.features.entropy` function computes the entropy of a histogram of the power
values measured in the spectrogram. The histogram represents an estimate of probability distribution function
of the power. You must build the histogram on your own, however. And you should also be
sure that your histogram is normalized to 1 (Sum h_i * bin_size_i  = 1). 

When used properly, this could score each spectrogram with a value between 0 and 1, where 1
represents pure noise and 0 would represent a maximally large signal, or amount of information.
(Though in reality, expect signals to have a value between 0.5 and 1, and for small signals to 
be closer to 1.0 than to 0.5.) 

The use of this measure depends significantly on how the histogram is created. There are 
multiple ways this can be done. See the [docstring for details on how to build a histogram and use this calculation.](ibmseti/features.py#L188-L282)

Example: 

```
bin_edges = range(0,501)
p, _ = np.histogram(np.clip(spectrogram.flatten(), 0, 500), bins=bin_edges, density=True)
w = np.diff(bin_edges)
h_p, h_max = ibmseti.features.entropy(p,w)

h_normal = h_p / h_max  #h_normal should range between 0 and 1.
```

##### Features based on the [First Difference](http://people.duke.edu/~rnau/411diff.htm)

###### Mean First Difference (along the time axis) of the spectrogram.

```
first_diff_along_time = ibmseti.features.difference(data, axis=0)
first_diff_along_freq = ibmseti.features.difference(data, axis=1)

fd_mean_along_time = np.mean(first_diff_along_time, axis=0).mean() 
fd_mean_along_freq = np.mean(first_diff_along_freq, axis=1).mean()

#same as first_diff_along_time.mean() because each element of the array is the same size!
#the average of the averages is the same as the total average if the sets of averages all have the same cardinality
```

###### N-th moment of the first difference

```
N = 2 #variation
N = 3 #kurtosis 
N = 4 #skewness
nth_moment_time = np.mean( ibmseti.features.moment( first_diff_along_time, axis=0, moment=N))
nth_moment_freq = np.mean( ibmseti.features.moment( first_diff_along_freq, axis=1, moment=N))
```

###### Maximum Variation

This can be calculated for the spectrogram, or for the first-difference, or gradient, 
along the time- or frequency-axis.

```
max_var_t = np.max( ibmseti.features.maximum_variation(ibmseti.features.difference(data, axis=0), axis=0))
max_var_f = np.max( ibmseti.features.maximum_variation(ibmseti.features.difference(data, axis=1), axis=1))
```


##### Features based on the Gradient

One can also calculate similar features (mean, N-th moments) based on the **gradient** of the signal along
the time or frequency axis. Use `ibmseti.features.first_order_gradient` 
or `numpy.gradient` to calculate the gradients. Then just as above with the
`first difference`, one can calculate the various moments and other features.

##### Asymmetry

It may be useful to examine the asymmetry of the signal measured in the L and R polarization components. 
The asymmetry is defined as

  A = (spect_L - spect_R) / (spect_L + spect_R)

where `spect_L` and `spect_R` are the 2D spectrograms for the L and R polarizations.  The returned 
`A` can be further analyzed. The total integration of `A` could be a useful feature as we would expect
E.T. signals not to be completely polarized in either the L or R polarizations (A = +/- 1).  

##### Summation of Polarities

It may be useful to examine the full signal in both polarities by summing them together

  S = spect_L + spect_R

where `spect_L` and `spect_R` are the 2D spectrograms for the L and R polarizations.  

### License 

Copyright 2017 IBM Cloud Data Services

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


### Contributing

Please feel free to fork this repository and submit Pull Requests. The core research team 
members from SETI, NASA and IBM will review your work and consider your algorithms for
inclusion in our library. We are hoping to find new algorithms that are useful enough to 
be included in future real-time analysis of data from the ATA and other SETI projects.





CHANGES.rst
2.0.1 (unreleased)
==================

2.0.0 (2018-03-23)
==================

- Released after testing

2.0.0.dev5 (2017-08-24)
==================

- [NEW] Support for Python 3. Tested to work with Python 2.7 and Python 3.6.
- [NEW] Uses six package to handle data input SimCompamp 

2.0.0.dev1 (2017-08-23)
==================

- [NEW] Support for Python 3. Tested to work with Python 2.7 and Python 3.6. 

1.0.5  (2017-06-21)
==================
- [NEW] Adding ibmseti.compamp.SimCompamp.sigProc function for inserting signal processing function before spectrogram calculations.

1.0.4 
==================

- [UPDATE] Automatically subtract any DC offset in Simulation data (compamp.SimCompamp)

1.0.3
==================
- [UPDATE] Changes default size of SimCompamp to 32 x 6144.

1.0.2 
==================
- [UPDATE] Moves SimCompamp class into compamp module instead of having separate file. :p

1.0.1
==================
- [NEW] Supports reading simulated data files prepared for the SETI Institute hackathon and code challenge, 2017.

1.0.0 (2017-03-21)
==================
- [NEW] Removes signaldb module, which wasn't so useful. Also makes the package independent of pyspark. 

0.0.8 (2016-11-07)
==================

- [NEW] Breaks out methods to compute signal in fourier space and to create a single 1D time-domain signal with over-sampled frequencies removed.

0.0.7 (2016-11-07)
==================

- [UPDATE] Compamp.complex_data returns a Numpy array of type complex64, reducing memory usage
- [NEW] Automatically imports ibmseti.constants (bins_per_half_frame is needed for DIY data manipulation)

0.0.6 (2016-10-19)
==================

- [NEW] Removes callback to track usage.

0.0.5 and earlier
===================
Changes were not tracked.

ibmseti/features.py
# Copyright (c) 2017 IBM. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

'''
This module contains some of the standard features that are extracted from
spectrograms and auto-correlation calculations from the raw SETI data.

Some functions are merely wrappers around Numpy-based operations, but
contain documentation that explicitly show how they are used with SETI data.

'''

import numpy as np
import scipy.stats
import math


def difference(arr, n=1, axis=0, **kwargs):
  '''
  Assuming that `arr` is a 2D spectrogram returned by
  ibmseti.dsp.raw_to_spectrogram(data), this function
  uses the Numpy.diff function to calculate the nth
  difference along either time or frequency.

  If axis = 0 and n=1, then the first difference is taken
  between subsequent time samples

  If axis = 1 and n=1, then the first difference is taken
  between frequency bins.

  For example:

    //each column is a frequency bin
    x = np.array([
       [ 1,  3,  6, 10],   //each row is a time sample
       [ 0,  5,  6,  8],
       [ 2,  6,  9, 12]])

    ibmseti.features.first_difference(x, axis=1)
    >>> array([[2, 3, 4],
               [5, 1, 2],
               [4, 3, 3]])

    ibmseti.features.first_difference(x, axis=0)
    >>> array([[-1,  2,  0, -2],
               [ 2,  1,  3,  4]])

  '''
  return np.diff(arr, n=n, axis=axis, **kwargs)


def projection(arr, axis=0, **kwargs):
  '''
  Assuming that `arr` is a 2D spectrogram returned by
  ibmseti.dsp.raw_to_spectrogram(data), where each row
  of the `arr` is a power spectrum at a particular time,
  this function uses the numpy.sum function to project the
  data onto the time or frequency axis into a 1D array.

  If axis = 0, then the projection is onto the frequency axis
  (the sum is along the time axis)

  If axis = 1, then the projection is onto the time axis.
  (the sum is along the frequency axis)

  For example:

    //each column is a frequency bin
    x = np.array([
       [ 1,  3,  6, 10],   //each row is a time sample
       [ 0,  5,  6,  8],
       [ 2,  6,  9, 12]])

    ibmseti.features.projection(x, axis=1)
    >>> array([20, 19, 29])

    ibmseti.features.projection(x, axis=0)
    >>> array([ 3, 14, 21, 30])

  One interesting kwarg that you may wish to use is `keepdims`.
  See the documentation on numpy.sum for more information.

  '''
  return np.sum(arr, axis=axis, **kwargs)


def moment(arr, moment=1, axis=0, **kwargs):
  '''
  Uses the scipy.stats.moment to calculate the Nth central
  moment about the mean.

  If `arr` is a 2D spectrogram returned by
  ibmseti.dsp.raw_to_spectrogram(data), where each row
  of the `arr` is a power spectrum at a particular time,
  this function, then the Nth moment along each axis
  will be computed.

  If axis = 0, then Nth moment for the data in each
  frequency bin will be computed. (The calculation is done
    *along* the 0th axis, which is the time axis.)

  If axis = 1, then Nth moment for the data in each
  time bin will be computed. (The calculation is done
    *along* the 1st axis, which is the frequency axis.)

  For example, consider the 2nd moment:

    //each column is a frequency bin
    x = array([[  1.,   3.,   6.,  10.], //each row is a time sample
               [  0.,   5.,   6.,   8.],
               [  2.,   6.,   9.,  12.]])

    ibmseti.features.mement(x, moment=2, axis=0) //the returned array is of size 4, the number of columns / frequency bins.
    >>>  array([ 0.66666667,  1.55555556,  2.,  2.66666667])

    ibmseti.features.mement(x, moment=2, axis=1) //the returned array is of size 3, the number of rows / time bins.
    >>>  array([ 11.5 ,  8.6875, 13.6875])

  If `arr` is a 1D array, such as what you'd get if you projected
  the spectrogram onto the time or frequency axis, then you must
  use axis=0.

  '''
  return scipy.stats.moment(arr, moment=moment, axis=axis, **kwargs)


def first_order_gradient(arr, axis=0):
  '''
  Returns the gradient of arr along a particular axis using
  the first order forward-difference.
  Additionally, the result is padded with 0 so that the
  returned array is the same shape as in input array.
  '''
  grad_arr = difference(arr, n=1, axis=axis)
  return np.insert(grad_arr, grad_arr.shape[axis], 0, axis=axis)


def total_variation(arr):
  '''
  If arr is a 2D array (N X M), assumes that arr is a spectrogram with time along axis=0.

  Calculates the 1D total variation in time for each frequency and returns an array
  of size M.

  If arr is a 1D array, calculates total variation and returns a scalar.

  Sum ( Abs(arr_i+1,j  - arr_ij) )

  If arr is a 2D array, it's common to take the mean of the resulting M-sized array
  to calculate a scalar feature.
  '''
  return np.sum(np.abs(np.diff(arr, axis=0)), axis=0)


def maximum_variation(arr):
  '''
  return np.max(arr, axis=0) - np.min(arr, axis=0)

  If `arr` is a 1D array, a scalar is returned.

  If `arr` is a 2D array (N x M), an array of length M is returned.
  '''
  return np.max(arr, axis=0) - np.min(arr, axis=0)


def tv_2d_isotropic(grad_0_arr, grad_1_arr):
  '''
  Calculates the Total Variation

  Assumes a 2D array.

  grad_0_arr is the gradient along the 0th axis of arr.
  grad_1_arr is the gradient along the 1st axis of arr.

  You can use the 1st order forward-difference measure
  of the gradient (the standard calculation). Or you
  can use the second_order central gradient.

  '''
  return np.sqrt(grad_0_arr**2 + grad_1_arr**2).sum()


def entropy(p, w):
  '''
  Computes the entropy for a discrete probability distribution function, as
  represented by a histogram, `p`, with bin sizes `w`,

   h_p = Sum -1 * p_i * ln(p_i / w_i)

  Also computes the maximum allowed entropy for a histogram with bin sizes `w`.

    h_max = ln( Sum w_i )

  and returns both as a tuple (h_p , h_max ). The entropy is in 'natural' units.

  Both `p` and `w` must be Numpy arrays.

  If `p` is normalized to 1 ( Sum p_i * w_i = 1), then
  the normalized entropy is equal toh_p / h_max and will
  be in the range [0, 1].

  For example, if `p` is a completely flat PDF (a uniform distribution), then
  the normalized entropy will equal 1, indicating maximum amount of disorder.
  (This is easily shown for the case where w_i = 1.)

  If the `p_i` is zero for all i except j and p_j = 1, then the entropy will be 0,
  indicating no disorder.

  One can use this entropy measurement to search for signals in the spectrogram.
  First we need to build a histogram of the measured power values in the spectrogram.
  This histogram represents an estimate of the probability distribution function of the
  observed power in the spectrogram.

  If the spectrogram is entirely noise, the resulting histogram should be quite flat and
  the normalized entropy ( h_p / h_max ) will approach 1. If there is a significant signal
  in the spectrogram, then the histogram will not be flat and the normalized entropy will
  be less than 1.

  The decision that needs to be made is the number of bins and the bin size. And unfortunately,
  the resulting entropy calculated will depend on the binning.

  Based on testing and interpretibility, we recommend to use a fixed number of bins that either
  span the full range of the power values in the spectrogram (0 to spectrogram.max()),
  or span a fixed range (for example, from 0 to 500).

  For example, you may set the range equal to the range of the values in the spectrogram.

    bin_edges = range(0,int(spectrogram.max()) + 2) #add 1 to round up, and one to set the right bin edge.
    p, _ = np.histogram(spectrogram.flatten(), bins=bin_edges, density=True)
    w = np.diff(bin_edges)
    h_p, h_max = ibmseti.features.entropy(p,w)

  If you choose to fix the range of the histogram, it is highly recommended that you use
  `numpy.clip` to ensure that any of the values in the spectrogram that are greater than
  your largest bin are not thrown away!

  For example, if you decide on a fixed range between 0 and 500, and your spectrogram
  contains a value of 777, the following code would produce a histogram where that 777 value
  is not present in the count.

    bin_edges = range(0,501)
    p, _ = np.histogram(spectrogram.flatten(), bins=bin_edges, density=True)
    w = np.diff(bin_edges)
    h_p, h_max = ibmseti.features.entropy(p,w)

  But if you clip the spectrogram, you can interpret the last bin as being "the number
  of spectrogram values equal to or greater than the lower bin edge".

    bin_edges = range(0,501)
    p, _ = np.histogram(np.clip(spectrogram.flatten(), 0, 500), bins=bin_edges, density=True)
    w = np.diff(bin_edges)
    h_p, h_max = ibmseti.features.entropy(p,w)

  You can also choose to fix the number of bins

    bins = 50
    p, bin_edges = np.histogram(spectrogram.flatten(), bins=bins, density=True)
    w = np.diff(bin_edges)
    h_p, h_max = ibmseti.features.entropy(p,w)

  It is suggested to use any of the following measures as features:

    bin range, spectrogram.min, spectrogram.max, number_of_bins, log(number_of_bins)
    entropy, max_entropy, normalized_entropy.

  Automatic Binning:

  While Numpy and AstroML offer ways of automatically binning the data, it is unclear if this
  is a good approach for entropy calculation -- especially when wishing to compare the value
  across different spectrogram. The automatic binning tends to remove disorder in
  the set of values, making the histogram smoother and more ordered than the data actually are.
  This is true of automatic binning with fixed sizes (such as with the 'rice', and 'fd' options in
  numpy.histogram), or with the variable sized arrays as can be calculated with Bayesian Blocks
  with astroML. However, nothing is ruled out. In preliminary testing,
  the calculated entropy from a histogram calculated with Bayesian Block binning seemed to be more
  sensitive to a simulated signal than using fixed binning. However, it's unclear how to
  interpret the results because "h_p/h_max" *increased* with the presence of a signal and exceeded 1.

  **It is likely that the calculation of h_max is done incorrectly. Please check my work!**

  It may even be that the total number of bins created by the Bayesian Block method would
  be a suitable feature. For a completely flat distribution, there will only be one bin. If the
  data contains significant variation in power levels, the Bayesian Block method will produce more
  bins.  More testing is required and your mileage may vary.

    import astroML.plotting

    bin_edges = astroML.density_estimation.bayesian_blocks(spectrogram.flatten())
    p, _ = np.histogram(spectrogram.flatten(), bins=bin_edges, density=True)
    w = np.diff(bin_edges)

    h_p, h_max = ibmseti.features.entropy(p,w)

  Also to note: Using astroML.density_estimation.bayesian_blocks takes prohibitively long!

  "Entropy" of raw data.

  If `p` is NOT a PDF, then you're on your own to interpret the results. In this case, you
  may set `w` = None and the calculation will assume w_i = 1 for all i.

  For example,

    h_p, _ = ibmseti.features.entropy(spectrogram.flatten(), None)

  '''
  if w is None:
    w = np.ones(len(p))

  h_p = np.sum([-x[0]*math.log(x[0]/x[1]) if x[0] else 0 for x in zip(p, w)])
  h_max = math.log(np.sum(w))

  return h_p, h_max


def asymmetry(spectrogram_L, spectrogram_R):
  '''
  returns (spectogram_L - spectrogram_R) / (spectogram_L + spectrogram_R)

  The asymmetry measure should be used for both polarizations recorded for a particular observation.

  On can then perform analysis such as integrating the returned spectrogram to determine if the
  asymmetry of the signals are close to zero, indicating equal signal in both polarizations, or
  close to +-1, indicating a strong asymmetry in the L or R polarization.

  spectrogram_L and spectrogram_R should be Numpy arrays.
  '''
  return (spectrogram_L - spectrogram_R) / (spectrogram_L + spectrogram_R)
